packages/commerce-sdk-react/src/hooks/ShopperCustomers/cache.ts

```diff
@@ -4,94 +4,253 @@
  */
+import {ShopperCustomersTypes} from 'commerce-sdk-isomorphic'
 import {Query} from '@tanstack/react-query'
-import {getCustomerProductListItem, QueryKeys} from './queryKeyHelpers'
-import {ApiClients, CacheUpdate, CacheUpdateMatrix, Tail} from '../types'
+import {getCustomerProductListItem} from './queryKeyHelpers'
+import {ApiClients, CacheUpdateMatrix} from '../types'
 import {
     getCustomer,
     getCustomerAddress,
     getCustomerPaymentInstrument,
     getCustomerProductList,
     getCustomerProductLists
 } from './queryKeyHelpers'
-import {and, pathStartsWith} from '../utils'
+import {and, clone, pathStartsWith} from '../utils'
 
 type Client = ApiClients['shopperCustomers']
+type Customer = ShopperCustomersTypes.Customer
+type CustomerProductList = ShopperCustomersTypes.CustomerProductList
+type CustomerProductListResult = ShopperCustomersTypes.CustomerProductListResult
 
 const noop = () => ({})
 
-/** Invalidates the customer endpoint, but not derivative endpoints. */
-const invalidateCustomer = (parameters: Tail<QueryKeys['getCustomer']>): CacheUpdate => ({
-    invalidate: [{queryKey: getCustomer.queryKey(parameters)}]
-})
+// ReactQuery type def for the updater function uses type `unknown` for the first param,
+// this makes it hard to define the `createUpdateFunction` below, so we created this type.
+type Updater<T> = (data: T | undefined) => T | undefined
+
+// This is a slight variation of above, but we know the data is not null.
+type DataProcessor<T> = (data: NonNullable<T>) => T
+
+/**
+ * Create an update handler for the cache matrix. This updater will implicitly clone and assert
+ * that the data being passed to is is not undefined.
+ *
+ * @param update
+ * @returns
+ */
+const createUpdateFunction =
+    <T>(update: DataProcessor<T>): Updater<T> =>
+    (data) =>
+        data ? update(clone(data)) : undefined
 
 export const cacheUpdateMatrix: CacheUpdateMatrix<Client> = {
     createCustomerAddress(customerId, {parameters}, response) {
         // getCustomerAddress uses `addressName` rather than `addressId`
         const newParams = {...parameters, addressName: response.addressId}
         return {
-            // TODO: Rather than invalidate, can we selectively update?
-            ...invalidateCustomer(newParams),
-            update: [{queryKey: getCustomerAddress.queryKey(newParams)}]
+            update: [
+                {
+                    queryKey: getCustomerAddress.queryKey(newParams)
+                },
+                {
+                    queryKey: getCustomer.queryKey(newParams),
+                    updater: createUpdateFunction((customer: Customer) => {
+                        // Push new address onto the end of addresses list.
+                        if (customer?.addresses) {
+                            customer.addresses = [...(customer.addresses ?? []), response]
+                        }
```

Comment:
If `addresses` is always present when a customer has addresses, then you shouldn't check for its existence. We know it's safe to add when it's missing but we have a new address.
```suggestion
                        customer.addresses = [...(customer.addresses ?? []), response]
```
---
If we only want to modify the `addresses` if they're already present, then your original solution was fine.
```suggestion
                        customer.addresses?.push(response)
```
(Mutating the array is fine because it's already been cloned.)

Reply:
I believe we'll still want to create an array in the scenario where the customers doesn't have an address array (this happens on new customers) So we could drop the condition and leave line 60